pymc_marketing/mmm/utils.py

```diff
@@ -33,3 +34,95 @@ def generate_fourier_modes(
+    The index of the point with the maximum distance to the line is returned as the elbow point.
+    """
+    # Calculate the coefficients of the line connecting the first and last points
+    line_coeffs = np.polyfit([x[0], x[-1]], [y[0], y[-1]], 1)
+
+    # Calculate the y-values of the line
+    line_y = np.poly1d(line_coeffs)(x)
+
+    # Calculate the distances from the points of the curve to the line
+    distances = distance.cdist(np.column_stack((x, y)), np.column_stack((x, line_y)), 'euclidean')
+
+    # Return the index of the point with the maximum distance to the line
+    return np.argmax(distances.diagonal())
+
+def calculate_curve(x: np.array, y: np.array) -> tuple[np.poly1d, np.ndarray, np.ndarray, np.ndarray, np.ndarray, list]:
+    """
+    Calculate the quadratic curve, its derivative, roots, and y values for given x values.
+
+    Parameters
+    ----------
+    x : array-like
+        The x-coordinates of the points.
+    y : array-like
+        The y-coordinates of the points.
+
+    Returns
+    -------
+    polynomial : numpy.poly1d
+        The quadratic polynomial representing the curve.
+    x_space_actual : numpy.ndarray
+        The x-values for the actual curve.
+    y_space_actual : numpy.ndarray
+        The y-values for the actual curve.
+    x_space_projected : numpy.ndarray
+        The x-values for the projected curve, including the roots.
+    y_space_projected : numpy.ndarray
+        The y-values for the projected curve, including the roots.
+    roots : list of float
+        The real roots of the derivative of the curve.
+
+    Notes
+    -----
+    This function fits a quadratic curve to the given points using the numpy.polyfit function.
+    It calculates the derivative of the curve using the numpy.poly1d.deriv method.
+    The function finds the real roots of the derivative using numpy.poly1d.r.
+    It defines the x-values for the actual curve using numpy.linspace with the minimum and maximum x values.
+    The x-values for the projected curve are defined using the minimum and maximum x values, including the roots.
+    The y-values for both curves are calculated using the polynomial function.
+
+    """
+
+    # Fit a quadratic curve
+    coefficients = np.polyfit(x, y, 2)
+    polynomial = np.poly1d(coefficients)
+
+    # Calculate derivative
+    derivative = polynomial.deriv()
+
+    # Find roots
+    roots = derivative.r
+    roots = [root.real for root in roots if root.imag == 0]
+
+    # Define x spaces
+    x_space_actual = np.linspace(x.min(), x.max(), 100)
+    x_space_projected = np.linspace(min(x.min(), min(roots)), max(x.max(), max(roots)), 100)
+
+    # Calculate y spaces
+    y_space_actual = polynomial(x_space_actual)
+    y_space_projected = polynomial(x_space_projected)
+
+    return polynomial, x_space_actual, y_space_actual, x_space_projected, y_space_projected, roots
```

Comment:
This function returns too many things. Maybe is worth using a `CurveResult` class ?